### PR TITLE
ENH: Volatility Control Charts

### LINKED ISSUE #32: ENH: control charts and volatility analysis
**Comments**
this would be a new visualizer. thanks to @antgonza for the suggestion.

**References**
See [compare_trajectories.py](http://qiime.org/scripts/compare_trajectories.html)

### LINKED ISSUE #35: input variable plot titles in HTML template
This is currently hardcoded (below is LME plot):
image

### PR DESCRIPTION
**Pull request by nbokulich**
fixes #35 

Well, it's a partial fix to #32, ~~but enough to close that issue~~. I added control charts, ~~but instead of using the qiime1 volatility analysis I use tests for equality of variance to examine volatility and data spread across time~~. LME, already in `q2-longitudinal`, appears to be a better way to examine differences in longitudinal trajectory than the qiime1 volatility methods. So I don't plan to implement the qiime1 methods ~~and we can close#32~~. 

**UPDATE 9/15/17:** the variance tests that I previously added assume sample independence, so are only appropriate for group comparisons at individual timepoints, not comparisons across time (which are the more important/interesting analysis). Hence we will only generate control charts for now, pending better stats. Still don't want to use the qiime1 volatility analyses, but will keep open that issue and document progress on the stats.

### CODE REVIEW

**Comment by nbokulich** [comment]
@thermokarst I just requested @gregcaporaso as a secondary reviewer — he and I had already discussed this visualizer a little bit; and since this is a whole new action I'm adding, it's a bit of a big PR. Thanks guys!

File: q2_longitudinal/tests/test_longitudinal.py [hunk 1]
```diff
@@ -115,6 +116,26 @@ def test_per_method_pairwise_stats_paired_nonparametric(self):
...
+            self.assertAlmostEqual(p, exp)
+            self.assertEqual(c, 6)
+
+    def test_compare_variances_unequal_variance(self):
+        for method, exp in zip(['fligner', 'levene', 'bartlett'],
+                               [0.2984133, 0.3024691, 0.00131798]):
+            groups = [[1, 2, 3], [4, 15, 96]]
+            s, p, c = compare_variances(*groups, method=method)
+            self.assertAlmostEqual(p, exp)
+
```
Comment (thermokarst) [inline q2_longitudinal/tests/test_longitudinal.py]: Can you add tests that check basic input/output testing of:

- `_add_sample_size_to_xtick_labels`
- `_control_chart_subplots`
- `_calculate_variability`
- `compare_variances`
- `_per_group_variance_comparison`
- `_multiple_tests_correction`
- `_control_chart`

Can you make sure that your tests reach into all of the branches in those functions? Thanks!

**Review by thermokarst — CHANGES_REQUESTED** [review]
Thanks @nbokulich --- I went through this from a user's perspective, so mostly just poking around the rendered viz, etc. The viz looks good, and I have no comments about the visualization params/signature! Code coverage looks good in abstract, but I would like to see some more discrete tests for all of the new utility functions --- these appear to be doing quite a bit of work, and it would be really nice to have that verified in the tests. Thanks, this will be really exciting to get in!

**Comment by nbokulich** [comment]
thanks @thermokarst ! I have added additional unit tests covering the listed functions (with the exception of `_control_chart` and `_control_chart_subplots`, in which the internal variance-calculating function is already tested). Please take a look and let me know if there are any more that you think will be required.

File: q2_longitudinal/tests/test_longitudinal.py [hunk 2]
```diff
@@ -115,6 +117,54 @@ def test_per_method_pairwise_stats_paired_nonparametric(self):
...
+    def test_add_sample_size_to_xtick_labels(self):
+        groups = {'a': [1, 2, 3], 'b': [1, 2], 'c': [1, 2, 3]}
+        labels = _add_sample_size_to_xtick_labels(groups)
+        self.assertEqual(labels, ['a (n=3)', 'b (n=2)', 'c (n=3)'])
+
+    def test_multiple_tests_correction(self):
```
Comment (thermokarst) [inline q2_longitudinal/tests/test_longitudinal.py:131]: Can you hit the `ZeroDivisionError` [exception](https://github.com/qiime2/q2-longitudinal/pull/36/files?diff=unified#diff-2c51f01d10d89d64b9690e69ef63e479R485) in this test, or another `multiple tests_correction` test?

File: q2_longitudinal/tests/test_longitudinal.py [hunk 3]
```diff
@@ -115,6 +117,54 @@ def test_per_method_pairwise_stats_paired_nonparametric(self):
...
+        test_df = pd.DataFrame(
+            pd.DataFrame({'Group': [1, 2, 3], 'P-value': [1, 1, 0.01]}))
+        test_df = _multiple_tests_correction(test_df)
+        self.assertEqual(
+            list(test_df['FDR P-value']), [1., 1., 0.030000000000000002])
+
+    def test_per_group_variance_comparison(self):
```
Comment (thermokarst) [inline q2_longitudinal/tests/test_longitudinal.py]: Can you add a test where `baseline is not None`? That way we can test [all the branches](https://github.com/qiime2/q2-longitudinal/pull/36/files?diff=unified#diff-2c51f01d10d89d64b9690e69ef63e479R448) in this function.

**Review by thermokarst — CHANGES_REQUESTED** [review]
Thanks @nbokulich --- a few minor additions.

**Comment by nbokulich** [comment]
Thanks @thermokarst ! I have added tests for both cases.

**Review by thermokarst — APPROVED** [review]
Thanks for the tests @nbokulich! I reviewed for vis structure, test coverage, and general python review. Time to send this over to @gregcaporaso for functionality review! Thanks!

**Comment by thermokarst** [comment]
LGTM, thanks @nbokulich. Passing this on to @gregcaporaso.

**Review by gregcaporaso — CHANGES_REQUESTED** [review]
Could you provide a description of what the dashed and dotted lines are in the control plots? That could possibly be a legend item in the plots. It looks like the solid line is the global mean, and the dotted and dashed lines are 2 and 3 standard deviations from the mean, respectively, but I'm not sure about that. 

It sounds like the code has been reviewed already and that you don't need me to do that. Is that correct?

**Comment by thermokarst** [comment]
@gregcaporaso: the code has been reviewed from a plugin perspective, but the data integrity has not been verified by way of review ("do the visualizations convey the same message as the stats?", "are the stats here meaningful to the questions we are attempting to answer?", etc.).

**Comment by gregcaporaso** [comment]
One more thought: Would it be worth adding a download link for the data used to generate these plots? Since we're not including any statistics, it could be useful to allow the user to get that data to do statistics on their own. If we did that, I think you'd want a tsv file that looks something like:

```
delivery  month  studyid  shannon
vaginal  0  42  2.2
cesarian  0  43  3.0
...
```

(**EDITED**: to make the example file tab-separated text instead of comma-separated)

**Comment by gregcaporaso** [comment]
> the data integrity has not been verified by way of review

Thanks @thermokarst, I've reviewed the viz from that perspective, and just had the two comments above.

**Comment by thermokarst** [comment]
@nbokulich: if you add a download link for the data, can you please provide that as a TSV at the very minimum? We are in the process of updating all visualizations to provide TSV as their primary download option (if they have a download option), and then if there are any other formats to download as, present those as secondary options (e.g. CSV, excel, etc.). Thanks!

**Comment by nbokulich** [comment]
@gregcaporaso your intuition about the lines is correct. This is described in the plugin description [here](https://github.com/qiime2/q2-longitudinal/pull/36/files#diff-2fd4a61578bada3af3efd746842f8bc6R180) and the docs tutorial (currently pending in [this PR](https://github.com/qiime2/docs/pull/200))

Do you think it is necessary to add descriptive text to the viz if this is described in the plugin description (so appears when the user seeks `--help`) and tutorial?

Downloading data should be easy to add. @thermokarst no worries, tsv is my preferred format as well.

**Comment by gregcaporaso** [comment]
> Do you think it is necessary to add descriptive text to the viz if this is described in the plugin description (so appears when the user seeks `--help`) and tutorial?

I think it'd be helpful, but not essential. Your call on this one @nbokulich.

**Comment by nbokulich** [comment]
@gregcaporaso could I add the data download in a separate PR? I probably ought to do this for all actions (while others do generate stats, raw data download may be useful)

**Comment by gregcaporaso** [comment]
Sure, that works. Want to create an issue for it and we can merge this one?

**Comment by nbokulich** [comment]
> Sure, that works. Want to create an issue for it and we can merge this one?

@gregcaporaso : #47

**Review by gregcaporaso — APPROVED** [review]
